sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,217 @@
+                    type == typeof(ETag) ||
+
+                    // Allowable for DynamicData assignments.
+                    type == typeof(JsonElement) ||
+                    type == typeof(JsonDocument) ||
+                    type == typeof(MutableJsonDocument) ||
+                    type == typeof(MutableJsonElement) ||
+                    type == typeof(DynamicData);
+            }
+
+            private static bool IsAllowedPrimitive(Type type)
+            {
+                return
+                    type.IsPrimitive ||
+                    type == typeof(decimal);
+            }
+
+            private static bool IsAllowedCollectionType<T>(Type type, T value)
+            {
+                return
+                    IsAllowedArrayType(type, value) ||
+                    IsAllowedListType(type, value) ||
+                    IsAllowedDictionaryType(type, value);
+            }
+
+            private static bool IsAllowedArrayType<T>(Type type, T value)
+            {
+                if (value is not Array array)
+                {
+                    return false;
+                }
+
+                Type? elementType = type.GetElementType();
+                if (elementType == null)
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(elementType))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(elementType, array);
+            }
+
+            private static bool IsAllowedListType<T>(Type type, T value)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (value is not IList list)
+                {
+                    return false;
+                }
+
+                Type elementType = type.GetGenericArguments()[0];
+                if (IsAllowedPrimitive(elementType))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(elementType, list);
+            }
+
+            private static bool IsAllowedDictionaryType<T>(Type type, T value)
+            {
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (value is not IDictionary dictionary)
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                if (types[0] != typeof(string))
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(types[1]))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(types[1], dictionary.Values);
+            }
+
+            private static bool IsAllowedEnumerableValue(Type elementType, IEnumerable enumerable)
```

Comment:
Can we use generic IEnumerable? This one will box values like DateTime

Reply:
I see. ... given we short circuit arrays or primitives (where boxing would be a big concern), I think we can box (and so use the non-generic IEnumerable) for the rest. Let's just make sure we don't run this routine for common priomitives.